x-pack/libbeat/management/generate.go

```diff
@@ -0,0 +1,253 @@
+	r.transformFunc = transform
+}
+
+// SetTransform sets a transform function callback
+func (r *TransformRegister) Transform(cfg *proto.UnitExpectedConfig, agentInfo *client.AgentInfo) ([]*reload.ConfigWithMeta, error) {
+	// If no transform is registered, fallback to a basic setup
+	if r.transformFunc == nil {
+		streamList, err := CreateInputsFromStreams(cfg, "log", agentInfo)
+		if err != nil {
+			return nil, fmt.Errorf("error creating input list from fallback function: %w", err)
+		}
+		// format for the reloadable list needed bythe cm.Reload() method
+		configList, err := CreateReloadConfigFromInputs(streamList)
+		if err != nil {
+			return nil, fmt.Errorf("error creating reloader config: %w", err)
+		}
+		return configList, nil
+	}
+
+	return r.transformFunc(cfg, agentInfo)
+}
+
+// ===========
+// Stream and Input processors
+// ===========
+
+// CreateInputsFromStreams breaks down the raw Expected config into an array of individual inputs/modules from the Streams values
+// that can later be formatted into the reloader's ConfigWithMetaData and sent to an indvidual beat/
+// This also performs the basic task of inserting module-level add_field processors into the inputs/modules.
+func CreateInputsFromStreams(raw *proto.UnitExpectedConfig, inputType string, agentInfo *client.AgentInfo) ([]map[string]interface{}, error) {
+	inputs := make([]map[string]interface{}, len(raw.Streams))
+
+	for iter := range raw.Streams {
+		streamSource := raw.Streams[iter].Source.AsMap()
+		streamSource = injectIndexStream(raw, inputType, iter, streamSource)
+		streamSource, err := injectStreamProcessors(raw, inputType, iter, streamSource)
+		if err != nil {
+			return nil, fmt.Errorf("Error injecting stream processors: %w", err)
+		}
+		streamSource, err = injectAgentInfoRule(streamSource, agentInfo)
+		if err != nil {
+			return nil, fmt.Errorf("Error injecting agent processors: %w", err)
+		}
+		inputs[iter] = streamSource
+	}
+
+	return inputs, nil
+}
+
+// CreateReloadConfigFromInputs turns a raw input/module list into the ConfigWithMeta type used by the reloader interface
+func CreateReloadConfigFromInputs(raw []map[string]interface{}) ([]*reload.ConfigWithMeta, error) {
+	// format for the reloadable list needed bythe cm.Reload() method
+	configList := make([]*reload.ConfigWithMeta, len(raw))
+
+	for iter := range raw {
+		uconfig, err := conf.NewConfigFrom(raw[iter])
+		if err != nil {
+			return nil, fmt.Errorf("error in conversion to conf.C: %w", err)
+		}
+		configList[iter] = &reload.ConfigWithMeta{Config: uconfig}
+	}
+	return configList, nil
+}
+
+// Emulates the InjectAgentInfoRule and InjectHeadersRule ast rules
+func injectAgentInfoRule(inputs map[string]interface{}, agentInfo *client.AgentInfo) (map[string]interface{}, error) {
+	// upstream API can sometimes return a nil agent info
+	if agentInfo == nil {
+		return inputs, nil
+	}
+	var processors = []interface{}{}
```

Comment:
https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

```suggestion
	var processors []interface{}
```

Or determine the required capacity and allocate that upfront.